yew-functional-macro/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+use quote::{format_ident, quote};
+use syn::parse::{Parse, ParseStream};
+use syn::{parse_macro_input, Block, FnArg, Ident, Item, Type};
+
+struct FunctionalComponent {
+    body: Box<Block>,
+    props_type: Type,
+    props_name: Ident,
+}
+
+impl Parse for FunctionalComponent {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let parsed: Item = input.parse()?;
+
+        match parsed {
+            Item::Fn(func) => {
+                if !func.sig.generics.params.is_empty() {
+                    // TODO maybe find a way to handle those
+                    return Err(syn::Error::new_spanned(
+                        func.sig.generics,
+                        "functional components cannot contain generics",
+                    ));
+                }
+
+                let inputs = &func.sig.inputs;
+
+                if inputs.len() > 1 {
+                    return Err(syn::Error::new_spanned(
+                        inputs,
+                        "functional components must take only parameter of props",
+                    ));
+                }
+
+                let (props_type, props_name) = if let Some(arg) = inputs.into_iter().next() {
+                    match arg {
+                        FnArg::Typed(arg) => {
+                            let ident = match &*arg.pat {
+                                syn::Pat::Ident(ident) => &ident.ident,
+                                pat => {
+                                    return Err(syn::Error::new_spanned(
+                                        pat,
+                                        "Cannot obtain ident. This should never happen",
+                                    ))
+                                }
+                            };
+
+                            let ty = match &*arg.ty {
+                                Type::Reference(ty) => {
+                                    if ty.lifetime.is_some() {
+                                        return Err(syn::Error::new_spanned(&ty.lifetime, "reference must not have life time"));
+                                    }
+
+                                    if ty.mutability.is_some() {
+                                        return Err(syn::Error::new_spanned(&ty.mutability, "reference must not be mutable"));
+                                    }
+
+                                    &*ty.elem
+                                }
+                                ty => return Err(syn::Error::new_spanned(ty, "invalid argument passed. (hint: type must be props struct by reference)"))
+                            };
+
+                            (ty.clone(), ident.clone())
+                        }
+
+                        arg => {
+                            return Err(syn::Error::new_spanned(
+                                arg,
+                                "functional components cannot accept a receiver",
+                            ));
+                        }
```

Comment:
We might as well make this an exhaustive match if there are only two options.
```suggestion
                        FnArg::Receiver(_) => {
                            return Err(syn::Error::new_spanned(
                                arg,
                                "functional components cannot accept a receiver",
                            ));
                        }
```